README.rst
rst server
===========

.. image:: https://travis-ci.org/unnonouno/rstserv.svg?branch=master
   :target: https://travis-ci.org/unnonouno/rstserv

This python script is a simple http server that shows a reStructured text file.

Require
-------

- python (> 2.4)

  - docutils
  - markdown (optional)


Setup
-----

::

 $ pip install rstserv


Usage
-----

::

$ rstserv [-p port] [file]

Run the script with your rst file, and access http://localhost:8080.


Command-line Options
~~~~~~~~~~~~~~~~~~~~

\-p
  port number (default: 8080)

License
-------

This program is distributred under the MIT license.


ChangeLog.rst
Release 0.1.0 - 2014/11/02
--------------------------

First release


test/function_test.py
#!/usr/bin/env python

import subprocess
import time
import os
import signal

proc = subprocess.Popen(['rstserv', 'README.rst'])
time.sleep(1)
ret = proc.poll()
if ret is not None:
    raise RuntimeError('cannot start server, return code is %d' % ret)

try:
    ret = subprocess.call(['curl', 'http://127.0.0.1:8080/'])
    if ret != 0:
        raise RuntimeError('curl returns %d' % ret)

finally:
    os.kill(proc.pid, signal.SIGKILL)


rstserv/core.py
import argparse
import imp
import os
import socket
import sys

from rstserv import __version__

if sys.version_info >= (3, 0):
    from http.server import BaseHTTPRequestHandler
    from http.server import HTTPServer
else:
    from BaseHTTPServer import BaseHTTPRequestHandler
    from BaseHTTPServer import HTTPServer

from docutils.core import publish_parts
from docutils.writers.html4css1 import Writer

try:
    import markdown
except:
    markdown = None

import rstserv

def read_all(path):
    f = open(path)
    try:
        return f.read()
    finally:
        f.close()


def rst2html(file_path):
    string = read_all(file_path)
    w = Writer()
    (_, path, _) = imp.find_module('template', rstserv.__path__)
    overrides = {
        'stylesheet': os.path.join(path, 'default.css'),
        'stylesheet_path': None,
        'initial_header_level': 2
        }
    r = publish_parts(string, writer=w, settings_overrides=overrides)
    result = r['whole']
    return result.encode('utf-8')


def md2html(file_path):
    source = read_all(file_path).decode('utf-8')
    body = markdown.markdown(source)
    html = '<html><body>\n' + body + '</body></html>\n'
    return body


def parse_args():
    p = argparse.ArgumentParser(
        description='rst server')
    p.add_argument('-p', '--port', default=8080,
                   type=int, help='port number')
    p.add_argument('file', metavar='FILE', help='rst file to show')
    return p.parse_args()


def main():
    args = parse_args()
    path = args.file

    class MyHandler(BaseHTTPRequestHandler):
        def do_GET(self):
            if markdown and path.endswith('.md'):
                html = md2html(path)
            else:
                html = rst2html(path)
            self.send_response(200)
            self.end_headers()
            self.wfile.write(html)
            return

    try:
        host = socket.gethostname()
        server = HTTPServer(('', args.port), MyHandler)
        print('rstserv %s' % __version__)
        print('Access http://%s:%i' % (host, args.port))
        print('Type <Ctrl-C> to stop the server')
        server.serve_forever()

    except Exception:
        e = sys.exc_info()[1]
        sys.stderr.write(str(e))
        sys.stderr.write('\n')
        sys.exit(1)


rstserv/__init__.py
'''
Copyright (c) 2011, Yuya Unno
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
    * Neither the name of the Yuya Unno nor the
      names of its contributors may be used to endorse or promote products
      derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
'''

import argparse
import imp
import os
import socket
import sys

if sys.version_info >= (3, 0):
    from http.server import BaseHTTPRequestHandler
    from http.server import HTTPServer
else:
    from BaseHTTPServer import BaseHTTPRequestHandler
    from BaseHTTPServer import HTTPServer

from docutils.core import publish_parts
from docutils.writers.html4css1 import Writer


def read_all(path):
    with open(path) as f:
        return f.read()


def rst2html(file_path):
    string = read_all(file_path)
    w = Writer()
    import rstserv
    (_, path, _) = imp.find_module('template', rstserv.__path__)
    overrides = {
        'stylesheet': os.path.join(path, 'default.css'),
        'stylesheet_path': None,
        'initial_header_level': 2
        }
    r = publish_parts(string, writer=w, settings_overrides=overrides)
    result = r['whole']
    return result.encode('utf-8')


def md2html(file_path):
    import markdown
    source = read_all(file_path).decode('utf-8')
    body = markdown.markdown(source)
    html = '<html><body>\n' + body + '</body></html>\n'
    return body


def parse_args():
    p = argparse.ArgumentParser(
        description='rst server')
    p.add_argument('-p', '--port', default=8080,
                   type=int, help='port number')
    p.add_argument('file', metavar='FILE', help='rst file to show')
    return p.parse_args()


def main():
    args = parse_args()
    path = args.file

    class MyHandler(BaseHTTPRequestHandler):
        def do_GET(self):
            if path.endswith('.md'):
                html = md2html(path)
            else:
                html = rst2html(path)
            self.send_response(200)
            self.end_headers()
            self.wfile.write(html)
            return

    host = socket.gethostname()
    server = HTTPServer(('', args.port), MyHandler)
    print('Access http://%s:%i' % (host, args.port))
    print('Type <Ctrl-C> to stop the server')
    server.serve_forever()
